walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```
@@ -13,11 +13,137 @@
+   * @param currentSlotIdx Index of element to be modified in {@queue}
+   */
+  private void combineTimeRange(List<TimeRange> queue, int currentSlotIdx) {
+    int adjSlotIdx = currentSlotIdx + 1;
+    TimeRange currentSlot = queue.get(currentSlotIdx);
+    TimeRange adjSlot = queue.get(adjSlotIdx);
+
+    // start combining adjacent/containing events with regards to current event
+    while (currentSlot.overlaps(adjSlot)) {
+      int curSlotStart = currentSlot.start();
+      int adjSlotEnd = adjSlot.end();
+      // Combine events if it is a chained event.
+      if (!currentSlot.contains(adjSlotEnd)) {
+        TimeRange combinedSlot = TimeRange.fromStartEnd(
+          curSlotStart, adjSlotEnd, false);
+          queue.set(currentSlotIdx, combinedSlot);
+      }
+      // Get next slot in line if available
+      queue.remove(adjSlotIdx);
+      if(adjSlotIdx == queue.size()) break;
+      else adjSlot = queue.get(adjSlotIdx);
+    }
+  }
+
+
+  /**
+   * Creates a vacant slot based on the provided index of the unavailable timeslot.
+   * 
+   * // |OE1|   <vacantEvent here>   |OE2|
+   * @param occupiedQueue Combined list of TimeRange with TimeRange that is unavailable
+   * @param slotIdx Index of TimeRange Object in front of the VacantSlot in {@occupiedQueue}
+   */
+  private TimeRange createVacantSlot(List<TimeRange> occupiedQueue, int slotIdx) {
+    TimeRange occupiedSlot1 = occupiedQueue.get(slotIdx);
+    TimeRange occupiedSlot2 = occupiedQueue.get(slotIdx + 1);
+    return TimeRange.fromStartEnd(occupiedSlot1.end(), occupiedSlot2.start(), false);
+  }
+
+  public Collection<TimeRange> query(Collection<Event> eventsSource, MeetingRequest request) {
+    List<TimeRange> vacantSlots= new ArrayList<TimeRange>();
+    List<TimeRange> occupiedQueue = new ArrayList<TimeRange>();
+    // Duplicate events list to support removal of events
+    List<Event> events = new ArrayList<>(eventsSource);
+
+    if (hasNoAttendees(request)) {
```

These two if-blocks can be executed first. The key is to fail fast without having to initialize the `List` above unnecessarily.